[Sung Jian Xun Joel] iP

Joel-Sung opened the pull request:
# My pull request for Duke
## https://github.com/nus-cs2103-AY2122S1/ip/pull/92

* Features added to Duke
  * Deadline, Event, Todo tasks
  * Add tasks
  * Delete tasks
  * Save tasks to memory
  * Find tasks by date/details

`<addr>` Reasons to pull:
1. Duke has more functions
1. UI looks ~~pretty~~ **MUCH** _prettier_

All you need to do is:
- [x] Accept my PR
- [ ] Download the src folder
- [ ] Run Duke.java

As a famous Duke once said:
> My Duke chat bot runs faster
> than yours.

A little teaser :wink: :
```javascript
public static void main(String[] args) {
        new Duke("taskList.txt").run();
}
```

Commit:
Add Gradle support

Joel-Sung committed:
Add greeting, echo, exit features to Duke

Joel-Sung committed:
Add list feature to Duke

Prints the text saved by Duke.

Joel-Sung committed:
Add Task.java and complete task feature to Duke

Duke now saves tasks and can indicate task completion.

Joel-Sung committed:
Add Deadline.java, Event.java, ToDo.java,

Created new types of Tasks.

Joel-Sung committed:
Add Duke exception classes and change exception handling in Duke

Makes exception handling unique to Duke and more concise.

Joel-Sung committed:
Add delete task feature to Duke

Joel-Sung committed:
Add text UI testing that checks Duke's output

Joel-Sung committed:
Edit Expected output

Joel-Sung committed:
Add packages

Organized files into packages.

Joel-Sung committed:
Add common function to add task to task list

Makes code easier to change.

Joel-Sung committed:
Add save feature to Duke

Task list is written to taskList.txt after Duke session is closed.

Edit text-ui-test to fit the new changes.

Joel-Sung committed:
Add date feature to Duke

Duke can now recognize dates.

Edit text-ui-test to test new features.

Joel-Sung committed:
Add date command to Duke to see which tasks are happening on that date

Edit text-ui-test to test for this feature.

Joel-Sung committed:
Add time feature to Duke

Duke is now able to recognize time. Add time to Deadline task and startTime and endTime to Event task.

Edit text-ui-test to test this feature.

Joel-Sung committed:
Merge branch 'branch-Level-7'

Joel-Sung committed:
Merge branch 'branch-Level-8' and 'branch-Level-7'

Edit code to resolve conflicts and ensure code from both branches work together.

Stop tracking of taskList.txt.

Joel-Sung committed:
Add more classes to handle the various features of Duke

Change runtest.bat to handle new classes.

Joel-Sung committed:
Add JUnit tests for Storage and TaskList

Joel-Sung committed:
Add META file to create JAR file

Fix some bugs.

Joel-Sung committed:
Add JavaDocs

Joel-Sung committed:
Tweak the code to comply with a coding standard

Joel-Sung committed:
Add find feature to Duke

Fix a bug where day values < 10 in date strings stored without preceding 0.

Edit text-ui-test to test find feature.

Joel-Sung committed:
Merge branch 'branch-A-JavaDoc'

Joel-Sung committed:
Merge branch branch-A-CodingStandard

Resolve file conflicts.

Joel-Sung committed:
Merge branch 'branch-Level-9'

Resolve file conflicts.

Add JavaDocs for find feature.

Joel-Sung committed:
Merge remote-tracking branch 'origin/add-gradle-support' into branch-A-Gradle

Joel-Sung committed:
Add gradle into project

Now able to run Duke and unit tests from Gradle.

Joel-Sung committed:
Merge remote-tracking branch 'origin/branch-A-Gradle'

Joel-Sung committed:
Add GUI using FXML to Duke, replace task date variable to LocalDateTime, add Pair class

Duke now returns output to MainWindow. MainWindow controls where the output is displayed in the GUI. Pair class helps Duke transfer data to MainWindow easier.

LocalDateTime is used to make parsing of user input easier. Now date format can only take yyyy-mm-dd. Previously dash could've been replaced with any other token.

Date feature no longer requires a time. It returns all tasks occurring at the given date. This way the command is more intuitive.

Joel-Sung committed:
Merge branch 'branch-A-GUI'

tanruiquan commented on src/main/java/Duke/gui/MainWindow.java:
```diff
@@ -0,0 +1,74 @@
+
+import duke.Duke;
+import duke.Pair;
+import javafx.fxml.FXML;
```
I'm really nitpicking here but perhaps adding a line break between `import duke.Pair` and `import javafx.fxml.FXML` would adhere better to the coding standards? I have notice this in several other places too, maybe something to consider?
```suggestion
import duke.Pair;

import javafx.fxml.FXML;
```

tanruiquan commented on src/main/java/Duke/task/Task.java:
```diff
@@ -0,0 +1,64 @@
+abstract public class Task {
+
+    private String details;
+    private boolean completed;
```
Maybe this boolean variable can be named `isCompleted` as well?
```suggestion
    private boolean isCompleted;
```

tanruiquan commented on src/main/java/Duke/gui/Main.java:
```diff
@@ -0,0 +1,73 @@
+import java.io.IOException;
+
+import duke.Duke;
+import javafx.application.Application;
```
I'm really nitpicking here but perhaps adding a line break between `import duke.Duke` and `import javafx.application.Application` would adhere better to the coding standards?
```suggestion
import duke.Duke;

import javafx.application.Application;
```

tanruiquan commented on src/main/java/Duke/exception/InvalidInputException.java:
```diff
@@ -0,0 +1,24 @@
+
+    public InvalidInputException(String str) {
+        super(str);
+        this.err = str;
```
Should `this.err` simply be `err` as it is not shadowed by a method or constructor parameter?
```suggestion
        err = str;
```

tanruiquan reviewed (commented):
Overall, I really like how clean your code is! It was super easy to read and understand. There were barely any coding standard violations. Keep up the good work! LGTM.

rehmmann commented on src/main/java/Duke/Duke.java:
```diff
@@ -71,148 +75,143 @@ private void saveTasks() throws IOException {
-                    contents += type + ' ' + done + ' ' + details + ' ' + at + ' '
-                            + dateStr + ' ' + start + ' ' + end;
-                }
+                    LocalDateTime timing = ((Event) task).getTiming();
```
Could be more descriptive in which timing is being referred to?

rehmmann commented on src/main/java/Duke/task/Event.java:
```diff
@@ -1,79 +1,75 @@
-    private int endTime;
+    private String timingStr;
+    private LocalDateTime timing;
+    private LocalTime endTime;
```
Might be difficult to differentiate these variables by their names?

rehmmann commented on src/main/java/Duke/ui/Ui.java:
```diff
@@ -4,143 +4,124 @@
-     * @param taskCount The current number of tasks in the task list.
+     * @return A message informing the task has been added.
+     */
+    public String addTaskMsg(Task task) {
+        return "Got it. I've added this task:\n"
+                + "\t" + task.toString() + "\n";
+    }
```
Good method name used here!Easy to understand its purpose :)

rehmmann commented on src/main/java/Duke/task/Event.java:
```diff
@@ -1,79 +1,75 @@
+                this.timing = null;
+                this.endTime = null;
+            }
+        } else {
+            timingStr = timing;
+            this.timing = null;
+            this.endTime = null;
```
Perhaps being consistent with using this to refer to class variables would be better?

rehmmann reviewed (commented):
Overall good naming conventions were seen and the code was easy to make sense of! LGTM

Joel-Sung committed:
Add CheckStyle

Format code to CheckStyle's requirements.

All switch statements now have default cases. Some methods now throw exception in the new default case. TaskList now has an additional constructor that takes in no arguments and initializes an empty ArrayList..................

Joel-Sung committed:
Merge branch 'branch-A-CheckStyle'

Joel-Sung committed:
Duke now creates storage file if it does not exist from the start

Previously, Duke only created the storage file upon exit and saving the tasks.

Joel-Sung committed:
Update shadowJar properties in build.gradle and ignore
META-INF/Manifest.MF files

Joel-Sung committed:
Use assertions to document important assumptions

There are certain areas in the code that assume some invariants. For
example in Duke.java, the saveTasks() method assumes task can only be
of type ToDo, Deadline or Event.

Assertions help to confirm these invariants. They will indicate that
there is something fundamentally wrong with the code, should any of
them be triggered.

Joel-Sung committed:
Improve code quality by abstracting code

Code was too messy and readability was bad. Abstracting code into
separate methods and classes made code easier to read and edit in the
future.

Storage exception handling for writing and reading is extracted from
Duke.java to Storage.java. StorageMissingException class created.

Input parsing for the various tasks is abstracted into common methods
and are now handled in their respective switch cases.

Each Task subclass now has a unique task label.

All formatters used for formatting date/time declared as constants.

Joel-Sung committed:
Add gradle.yml for Continuous Integration (CI)

Joel-Sung committed:
Merge pull request #1 from Joel-Sung/branch-A-Assertions

Add assertions to document important assumptions

Joel-Sung committed:
Merge branch 'master' of https://github.com/Joel-Sung/ip

Joel-Sung committed:
Merge branch 'branch-A-CodeQuality'

Resolved Conflicts in:
	src/main/java/Duke/Duke.java
	src/main/java/Duke/parser/Parser.java
	src/main/java/Duke/storage/Storage.java

Joel-Sung committed:
Merge branch 'branch-A-CI'

Joel-Sung committed:
Add sort feature to Duke

It is helpful to see tasks in chronological order so that users will
know which tasks are more urgent.

Sort allows user to list the tasks by date.
All tasks that are completed are listed at the start of the list.
Sort reverse allows user to reverse the list.

Deadline and Event no longer accommodate for incorrect format of date
time inputs. An exception is now thrown and the user is informed of the
incorrect format.

Storing incorrect date time inputs as strings makes the sorting very
confusing.

Joel-Sung committed:
Merge pull request #3 from Joel-Sung/branch-C-Sort

Add sort feature to Duke

Joel-Sung committed:
Merge branch 'master' of https://github.com/Joel-Sung/ip

Joel-Sung committed:
Improve GUI to make Duke look better

Duke's GUI needed to look more pleasing and polished.

Changes:
1. Spacing between dialog boxes
2. Dialog boxes are color coded based on user(blue),
Duke(green) and error messages(red)
3. Duke and user icon images made into a circle
4. Made task list font better
5. Added borders to input box and send button
6. Dialog boxes now expand upwards on very long messages

Combined exception messages in UI into one method as the exception
methods were very similar.

Created DukeResponse class to replace Pair class as GUI now needed 3
pieces of information from Duke.

Fix a bug where Duke cannot retrieve task details with spacing. Tasks
stored in storage now have their details and timing on separate lines.

Joel-Sung committed:
Rename Duke to Dog-and-Cat (DaC)

Give my own personality to the chatbot.

Joel-Sung committed:
Update README.md

Changed title of GUI from Duke to Dog-and-Cat (DaC)

Joel-Sung committed:
Update README.md

Joel-Sung committed:
Prepare for final release

Updated README files to include instructions for installation and user
guide.

Changed root package from duke to dac.

Set all font in GUI to Arial to support cross-platform.

Updated JUNIT test files. Removed text-ui-test folder.

Extracted some code into constants for better readability in
DialogBox.java.

Joel-Sung committed:
Merge branch 'master' of https://github.com/Joel-Sung/ip

# Conflicts:
#	docs/README.md

Joel-Sung committed:
Update README.md

Joel-Sung committed:
Update README.md

Joel-Sung committed:
Remove IO redirection tests from CI

Rename completed variable to amCompleted in Task.java

Joel-Sung committed:
Merge branch 'master' of https://github.com/Joel-Sung/ip

Joel-Sung committed:
Fix JAR file not saving on MAC OS

Joel-Sung committed:
Fix boolean variable naming convention
